feat: add a Message Subscription Process Definition Statistics API (ES/OS implementations)

johnBgood opened the pull request:
## Description

This pull request introduces improvements and new features to the search client, focusing on more flexible top hits aggregation, consistent field value handling, and support for message subscription process definition statistics. The main changes include refactoring the `SearchTopHitsAggregator` to support multiple fields and custom sorting, standardizing field value conversion logic across Elasticsearch and OpenSearch transformers, and adding new support for message subscription process definition statistics queries and their associated transformers.

### Top Hits Aggregator Enhancements

* Refactored `SearchTopHitsAggregator` to accept a list of fields (`fields`) and a `SortOption` instead of a single field and fixed sort order, allowing for more flexible queries and sorting. Updated builder methods and default values accordingly. [[1]](diffhunk://#diff-054d778f5b48af8a1e08db68f18962ebe376e0d6bcc0b0840fdf5dcfc6972646R10-R19) [[2]](diffhunk://#diff-054d778f5b48af8a1e08db68f18962ebe376e0d6bcc0b0840fdf5dcfc6972646L35-R39) [[3]](diffhunk://#diff-054d778f5b48af8a1e08db68f18962ebe376e0d6bcc0b0840fdf5dcfc6972646L45-R59) [[4]](diffhunk://#diff-054d778f5b48af8a1e08db68f18962ebe376e0d6bcc0b0840fdf5dcfc6972646L66-R85)
* Updated the Elasticsearch and OpenSearch top hits aggregator transformers to use the new `fields` and `SortOption` for source filtering and sorting, supporting multiple fields and custom sort orders. [[1]](diffhunk://#diff-96bc7a157dfd75adf626c27320c8af6dff60cb8fb1466de3d1cdbc56f2c61a16L30-R51) [[2]](diffhunk://#diff-c4956411053c8c1e4dfdbbca8d46c1bddc42a2ce660efe1d4668c837420b0f7aL30-R51)

### Field Value Conversion Consistency

* Introduced a helper method (`fieldValueToString`) in both Elasticsearch and OpenSearch aggregation result transformers to standardize conversion of field values to strings, ensuring consistency in composite key handling and pagination logic. [[1]](diffhunk://#diff-94c46f968e4dd355de4ecb06fe2856786e47cd2ba3bfc7d41a8b8df98d5de5bbL148-R148) [[2]](diffhunk://#diff-94c46f968e4dd355de4ecb06fe2856786e47cd2ba3bfc7d41a8b8df98d5de5bbL170-R183) [[3]](diffhunk://#diff-c4991e7b556bf18f2f29858e87fd27a635491dea74c5ebb0d54d8ca7ad041062L147-R147) [[4]](diffhunk://#diff-c4991e7b556bf18f2f29858e87fd27a635491dea74c5ebb0d54d8ca7ad041062L170-R182)

### Message Subscription Process Definition Statistics Support

* Added a new document reader for message subscription process definition statistics queries (`MessageSubscriptionProcessDefinitionStatisticsDocumentReader`), enabling aggregation and result handling for these statistics.
* Registered new aggregation, result, filter, query, and sort transformers for message subscription process definition statistics in the service transformers, integrating this feature into the search client infrastructure. [[1]](diffhunk://#diff-f562c57454f033f69b55e54a83cdb910e5f1a808b3f70f15c2a8f78638d4e510R11-R19) [[2]](diffhunk://#diff-f562c57454f033f69b55e54a83cdb910e5f1a808b3f70f15c2a8f78638d4e510R31-R39) [[3]](diffhunk://#diff-f562c57454f033f69b55e54a83cdb910e5f1a808b3f70f15c2a8f78638d4e510L79-R87) [[4]](diffhunk://#diff-f562c57454f033f69b55e54a83cdb910e5f1a808b3f70f15c2a8f78638d4e510L114-R122) [[5]](diffhunk://#diff-f562c57454f033f69b55e54a83cdb910e5f1a808b3f70f15c2a8f78638d4e510L141-R149) [[6]](diffhunk://#diff-f562c57454f033f69b55e54a83cdb910e5f1a808b3f70f15c2a8f78638d4e510R179) [[7]](diffhunk://#diff-f562c57454f033f69b55e54a83cdb910e5f1a808b3f70f15c2a8f78638d4e510L204-R213) [[8]](diffhunk://#diff-f562c57454f033f69b55e54a83cdb910e5f1a808b3f70f15c2a8f78638d4e510R357)

## Checklist

<!--- Please delete options that are not relevant. Boxes should be checked by reviewer. -->
- [ ] Enable backports when necessary (fex. [for bug fixes](https://github.com/camunda/camunda/blob/main/CONTRIBUTING.md#backporting-changes) or [for CI changes](https://github.com/camunda/camunda/wiki/CI-&-Automation#when-to-backport-ci-changes)).

## Related issues

closes #39613

johnBgood committed:
feat: add a Message Subscription Process Definition Statistics API (ES/OS implementations)

johnBgood committed:
refactor: fix variable name and typos

tmetzke commented on zeebe/gateway-protocol/src/main/proto/rest-api.yaml:
```diff
@@ -4475,6 +4475,37 @@ paths:
+      operationId: getMessageSubscriptionProcessDefinitionStatistics
+      summary: Get message subscription statistics for process definitions
+      description: |
+        Get message subscription statistics for process definitions based on given criteria.
```
```suggestion
        Get message subscription statistics, grouped by process definition.
```

tmetzke commented on zeebe/gateway-protocol/src/main/proto/rest-api.yaml:
```diff
@@ -4475,6 +4475,37 @@ paths:
+        content:
+          application/json:
+            schema:
+              $ref: "#/components/schemas/MessageSubscriptionProcessDefinitionStatisticsQuery"
```
🔧 We should switch the entities in the component names.
```suggestion
              $ref: "#/components/schemas/ProcessDefinitionMessageSubscriptionStatisticsQuery"
```

tmetzke commented on zeebe/gateway-protocol/src/main/proto/rest-api.yaml:
```diff
@@ -4475,6 +4475,37 @@ paths:
+              $ref: "#/components/schemas/MessageSubscriptionProcessDefinitionStatisticsQuery"
+      responses:
+        "200":
+          description: The message subscription statistics for process definitions result.
```
```suggestion
          description: The process definition message subscription statistics result.
```

tmetzke reviewed (commented):
👍 Looks really good to me already
💭 Based on the [comment](https://github.com/camunda/camunda/issues/39613#issuecomment-3455637934) in the issue, I have made some suggestions (they are not complete, just hinting at a pattern of change we'd need).
⚠️ I didn't get a chance to look at the statistics aggregation code yet. In the end, proper test cases are the most important thing to ensure that different use cases work as expected for those statistics, I guess.

tmetzke commented on search/search-client-elasticsearch/src/main/java/io/camunda/search/es/transformers/aggregator/SearchAggregationResultTransformer.java:
```diff
@@ -167,14 +167,20 @@ private <B extends MultiBucketBase> Object[] extractSearchAfter(
-              .collect(Collectors.toMap(Map.Entry::getKey, entry -> entry.getValue().stringValue()))
+              .collect(
+                  Collectors.toMap(
+                      Map.Entry::getKey, entry -> fieldValueToString(entry.getValue())))
```
Just trying to understand why we didn't run into this yet.

johnBgood commented on zeebe/gateway-protocol/src/main/proto/rest-api.yaml:
```diff
@@ -4475,6 +4475,37 @@ paths:
           $ref: "#/components/responses/InternalServerError"
       x-eventually-consistent: true
+
+  /message-subscriptions/statistics/process-definitions:
```
I wasn't sure indeed, but it makes sense. I'll update accordingly

johnBgood committed:
feat: add Process Def Message Subscription Statistics to the Camunda client + code review comments

# Conflicts:
#	dist/src/main/java/io/camunda/application/commons/rdbms/RdbmsConfiguration.java
#	dist/src/main/java/io/camunda/application/commons/search/SearchClientDatabaseConfiguration.java
#	dist/src/main/java/io/camunda/application/commons/search/SearchClientReaderConfiguration.java
#	search/search-client-query-transformer/src/main/java/io/camunda/search/clients/transformers/ServiceTransformers.java
#	search/search-client/src/main/java/io/camunda/search/clients/CamundaSearchClients.java
#	search/search-client/src/main/java/io/camunda/search/clients/ProcessDefinitionSearchClient.java
#	search/search-client/src/main/java/io/camunda/search/clients/impl/NoDBSearchClientsProxy.java
#	search/search-client/src/main/java/io/camunda/search/clients/impl/NoopSearchClientsProxy.java
#	service/src/main/java/io/camunda/service/ProcessDefinitionServices.java
#	zeebe/gateway-protocol/src/main/proto/rest-api.yaml
#	zeebe/gateway-rest/src/main/java/io/camunda/zeebe/gateway/rest/controller/ProcessDefinitionController.java
#	zeebe/gateway-rest/src/main/java/io/camunda/zeebe/gateway/rest/mapper/search/SearchQueryResponseMapper.java

johnBgood committed:
feat: rename EventEntity to MessageSubscriptionEntity in the process definition message subscription statistics API

tmetzke commented on zeebe/gateway-protocol/src/main/proto/rest-api.yaml:
```diff
@@ -8421,6 +8452,44 @@ components:
+        filter:
+          description: The message subscription filters.
+          allOf:
+            - $ref: "#/components/schemas/MessageSubscriptionFilter"
```
❓ Do we already need a filter here? If we do, is message subscriptions the right one? If we don't know yet, we can also leave it out for now and add it later when needed.

johnBgood replied:
It should work by default, I guess this filter is quite useful as you might request statistics for a specific state, process Id, or even process definition key. 
I'd advocate for keeping the filter here, I'm pretty we(connectors) will need it at least.

tmetzke commented on zeebe/gateway-protocol/src/main/proto/rest-api.yaml:
```diff
@@ -4512,6 +4512,37 @@ paths:
+
+  /process-definitions/statistics/message-subscriptions:
+    post:
+      tags:
```
🔧 We need to define the consistency behavior for all paths.

```suggestion
      x-eventually-consistent: true
      tags:
```

tmetzke commented on zeebe/gateway-rest/src/main/java/io/camunda/zeebe/gateway/rest/controller/ProcessDefinitionController.java:
```diff
@@ -134,6 +136,19 @@ public ResponseEntity<ProcessDefinitionElementStatisticsQueryResult> elementStat
+      processDefinitionStatistics(
+          @RequestBody(required = false)
+              final ProcessDefinitionMessageSubscriptionStatisticsQuery searchRequest) {
+    return SearchQueryRequestMapper.toMessageSubscriptionProcessDefinitionStatisticsQuery(
```
```suggestion
    return SearchQueryRequestMapper.toProcessDefinitionMessageSubscriptionStatisticsQuery(
```

tmetzke commented on zeebe/gateway-rest/src/main/java/io/camunda/zeebe/gateway/rest/mapper/search/SearchQueryRequestMapper.java:
```diff
@@ -557,6 +557,24 @@ public static Either<ProblemDetail, MessageSubscriptionQuery> toMessageSubscript
+          final ProcessDefinitionMessageSubscriptionStatisticsQuery request) {
+    if (request == null) {
+      return Either.right(
+          SearchQueryBuilders.messageSubscriptionProcessDefinitionStatisticsQuery().build());
```
```suggestion
          SearchQueryBuilders.processDefinitionMessageSubscriptionStatisticsQuery().build());
```

tmetzke commented on zeebe/gateway-rest/src/main/java/io/camunda/zeebe/gateway/rest/mapper/search/SearchQueryResponseMapper.java:
```diff
@@ -1217,5 +1220,34 @@ public static ProcessInstanceCallHierarchyEntry toProcessInstanceCallHierarchyEn
   }
 
+  private static List<ProcessDefinitionMessageSubscriptionStatisticsResult>
+      toMessageSubscriptionProcessDefinitionStatisticsQueryResponse(
```
```suggestion
      toProcessDefinitionMessageSubscriptionStatisticsQueryResponse(
```

tmetzke commented on qa/acceptance-tests/src/test/java/io/camunda/it/client/ProcessDefinitionStatisticsTest.java:
```diff
@@ -971,4 +976,62 @@ private static UserTask getUserTask(final long processDefinitionKey) {
+              final var subscriptions =
+                  camundaClient.newMessageSubscriptionSearchRequest().send().join();
+              assertThat(subscriptions.items()).hasSize(2);
+            });
```
❓ Am I missing something is this a duplicated wait condition? 😅

tmetzke reviewed (commented):
👍 Thanks for the adjustments you made 🙏 
💭 I left a couple of thoughts, nothing major, I would say.
🔧 We should add more test cases to ensure this does what we expect in more complex scenarios.

johnBgood committed:
refactor: rename all methods for the PD Message subscription statistics endpoint

# Conflicts:
#	dist/src/main/java/io/camunda/application/commons/rdbms/RdbmsConfiguration.java

johnBgood committed:
feat: add PD Message Subscriptions statistics endpoint in the main file

johnBgood committed:
fix: changes after rebasing onto main

johnBgood committed:
fix: RDBMS Process Def Message Subscription statistics + add tests

# Conflicts:
#	qa/acceptance-tests/src/test/java/io/camunda/it/client/ProcessDefinitionStatisticsIT.java

johnBgood committed:
fix: Extract sort to highlight that we don't support custom sorting

johnBgood committed:
fix: add missing @RequiresSecondaryStorage in ProcessDefinitionController

johnBgood committed:
test: add ProcessDefinitionStatisticsMultiTenantsTest

johnBgood committed:
fix: use the new ResourceAccessChecksg#etAuthorizedResourceIdsByType method

johnBgood committed:
fix: checkstyles violations

johnBgood committed:
test: fix unit tests for TopHits aggregation

johnBgood committed:
test: fix unit tests for TopHits aggregations is OS + apply spotless

johnBgood committed:
test: fix checkstyle issue

tmetzke reviewed (commented):
Some in-between review feedback, mostly regarding API and client code so far.
For future features, you can also think about breaking this up a bit and creating a PR with API and Java SDK first. That way, it's clear what we want to support externally. That PR can return empty results from the service layer, for example. The data layer-intensive PR can then follow up.

That way, we can clarify the API and other details upfront, and changes to that PR stay comparatively small, if needed. Also allows for faster feedback cycles.

johnBgood commented on db/rdbms/src/main/java/io/camunda/db/rdbms/read/service/ProcessDefinitionMessageSubscriptionStatisticsDbReader.java:
```diff
@@ -0,0 +1,74 @@
+  // we don't allow sorting by other fields for this statistics type
+  private static final ProcessDefinitionMessageSubscriptionStatisticsSort FIXED_SORT =
+      ProcessDefinitionMessageSubscriptionStatisticsSort.of(
+          b -> b.processDefinitionKey().asc().tenantId().asc());
```
We don't offer sorting (in this version at least), so this is the sorting used in the aggregations. I just wanted to make it explicit, and IIRC it was not sorted by default in RDBMS (only in ES)

johnBgood commented on zeebe/gateway-protocol/src/main/proto/v2/process-definitions.yaml:
```diff
@@ -432,6 +463,53 @@ components:
+        page:
+          description: The pagination information, limited to 'from' and 'limit'. Backward pagination is not supported.
+          allOf:
+            - $ref: "search-models.yaml#/components/schemas/SearchQueryPageRequest"
```
no, only `limit` and `after` in this iteration

tmetzke reviewed (commented):
👍 This looks amazing already, really nice job, @johnBgood 🧑‍🚀 
👍 The core logic and API are done from my point of view. The acceptance tests are really helpful here to validate that everything works as expected 🌟 
💭 I left some smaller items to look into, nothing major.

johnBgood committed:
refactor: ProcessDefinitionMessageSubscriptionStatistics returns a page instead of a endCursor

johnBgood committed:
refactor: ProcessDefinitionMessageSubscriptionStatistics use CursorForwardPagination

tmetzke reviewed (approved):
👍 Adjustments look good to me 🚀 
🔧 Something's off with the code, probably a compilation issue? I haven't looked at the CI.

johnBgood committed:
refactor: ProcessDefinitionMessageSubscriptionStatistics use CursorForwardPagination

johnBgood committed:
refactor: use TestStandaloneBroker instead of TestCamundaApplication
